flytekit/extras/pytorch/checkpoint.py

```diff
@@ -0,0 +1,141 @@
+import pathlib
+import typing
+from dataclasses import asdict, dataclass, fields, is_dataclass
+from typing import Any, Callable, Dict, NamedTuple, Optional, Type, Union
+
+import torch
+from dataclasses_json import dataclass_json
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+
+try:
+    from typing import Protocol
+except ImportError:
+    from typing_extensions import Protocol
+
+
+class IsDataclass(Protocol):
+    __dataclass_fields__: Dict
+    __dataclass_params__: Dict
+    __post_init__: Optional[Callable]
```

Comment:
just for my own understanding, why does this work to detect if an object is a dataclass?

Reply:
Every dataclass has `__dataclass_fields__` and `__dataclass_params__` defined. So this class is to define the type for reference. It doesn't really verify the type of the object.  That's handled by `__post_init__`.

Reply:
Nice, thanks for explaining. Is this documented anywhere? I just want to make sure this check is not a one-off and it's actually the way people usually type-check dataclasses.

Reply:
I don't think there's any official documentation. I came across this notation on [stackoverflow](https://stackoverflow.com/questions/54668000/type-hint-for-an-instance-of-a-non-specific-dataclass). If there's a better way to type-hint dataclass, let me know!